### PR TITLE
Cross-origin iframes and feature policy.

### LINKED ISSUE #281: Block payment handler in iframes by default
Similar to Payment Request, let's block Payment Handler APIs in cross-origin iframes by default. We can use [Feature Policy](https://wicg.github.io/feature-policy/) to selectively enable it like so:

```html
<iframe src="https://other.com/paymenthandler" allow="payment"></iframe>
```

The `allow="payment"` attribute also enables Payment Request.

### PR DESCRIPTION
**Pull request by rsolomakhin**
Prohibit cross-origin iframes from installing payment handlers, unless
allowed by the feature policy.

closes #281

The following tasks have been completed:

 * [ ] web platform tests
 * [ ] MDN Docs added

Implementation commitment:

 * [ ] Safari
 * [x] [Chrome](https://crbug.com/828948)
 * [ ] Firefox
 * [ ] Edge

***
<a href="https://pr-preview.s3.amazonaws.com/rsolomakhin/payment-handler/pull/282.html" title="Last updated on Sep 19, 2018, 6:53 PM GMT (eb1fb88)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/payment-handler/282/697480d...rsolomakhin:eb1fb88.html" title="Last updated on Sep 19, 2018, 6:53 PM GMT (eb1fb88)">Diff</a>

### CODE REVIEW

**Comment by rsolomakhin** [comment]
@marcoscaceres: Am I holding it right?

**Review by ghost — APPROVED** [review]

File: index.html [hunk 1]
```diff
@@ -2078,6 +2097,17 @@ <h2>
           </li>
         </ul>
       </section>
+      <section>
+        <h2>
+          Iframes
+        </h2>
+        <ul>
+          <li>Cross-origin iframes should not be able to register payment
```
Comment (marcoscaceres) [inline index.html]: "should not" doesn't read well here, and neither does "iframes". Perhaps just state that:

> The top-level document needs to explicitly grant access to nested browsing contexts via the "payment" feature policy. 

I've filed https://github.com/WICG/feature-policy/issues/154 so we can actually link to "payment".

File: index.html [hunk 2]
```diff
@@ -426,6 +426,10 @@ <h2>
             <li>Return <var>p</var> and perform the remaining steps in
             parallel:
             </li>
+            <li>If the <a>document</a> is not <a>allowed to use</a> the
```
Comment (marcoscaceres) [inline index.html]: Nit: drop "policy-controlled-feature", it's redundant.

File: index.html [hunk 3]
```diff
@@ -426,6 +426,10 @@ <h2>
...
+            policy-controlled-feature <code>payment</code>, reject <var>p</var>
+            with <a>SecurityError</a>.
```
Comment (marcoscaceres) [inline index.html]: Here, you want:

```HTML
"<a>SecurityError</a>" <a>DOMException</a>.
```

You can take the `dfn`s from payment request.

**Review by marcoscaceres — CHANGES_REQUESTED** [review]
Couple of nits, but generally looks great.

File: index.html (hunk 1 above)
Comment (marcoscaceres) [reply index.html]: Fixed wording above.

**Comment by rsolomakhin** [comment]
@marcoscaceres Please take another look! I've taken your suggestions.

**Comment by marcoscaceres** [comment]
Generally, this looks good. 

Nit: I think you can probably drop all the "policy-controlled-feature"s throughout.

**Comment by rsolomakhin** [comment]
@marcoscaceres: I've applied your fixes throughout. OK to merge?

**Comment by ianbjacobs** [comment]
Hi @rsolomakhin ,

Are we ready to merge this pull request?

Ian

**Comment by rsolomakhin** [comment]
@ianbjacobs: I'm ready to go, just need a review from @marcoscaceres, I think.

**Review by marcoscaceres — APPROVED** [review]

**Comment by marcoscaceres** [comment]
So, editorially it looks fine. Any plans to finish the checklist above before merging tho? I assume there are tests already.

**Comment by marcoscaceres** [comment]
I wonder if we should  get a second pair of eyes on the Feature Policy stuff before we merge? I don't know that spec very well, and having assurance from one of the editors of that spec that we've not missed anything might be good. Just a thought.

**Comment by rsolomakhin** [comment]
I will work on more web platform tests eventually, but not this week. We can wait for more web platform tests from Chrome to land before merging this in.

**Comment by rsolomakhin** [comment]
Now that PaymentInstruments API is removed, the only way to install a payment handler is during a transaction ("just-in-time") through `PaymentRequest.show()`, which already has the desirable properties with iframes and feature policy. Therefore, this pull request can be safely abandoned.

**Pull request closed without merging**